pageserver/src/walreceiver.rs

```diff
@@ -648,68 +668,78 @@ impl WalConnectionManager {
                     }
                 }
             })
-            .max_by_key(|(_, info, _)| info.commit_lsn)?;
-
-        match self.wal_connection_data.as_ref() {
-            None => Some(NewWalConnectionCandidate {
-                safekeeper_id: new_sk_id,
-                wal_producer_connstr: new_wal_producer_connstr,
-                reason: ReconnectReason::NoExistingConnection,
-            }),
-            Some(current_connection) => {
-                if current_connection.safekeeper_id == new_sk_id {
-                    None
-                } else {
-                    self.reason_to_reconnect(current_connection, new_sk_timeline)
-                        .map(|reason| NewWalConnectionCandidate {
-                            safekeeper_id: new_sk_id,
-                            wal_producer_connstr: new_wal_producer_connstr,
-                            reason,
-                        })
+            .max_by_key(|(_, info, _)| info.commit_lsn);
+
+        match (current_sk_data_updated, candidate_sk_data) {
+            // No better candidate than one we're already connected to:
+            // whatever data update comes for the connected one, we don't have a better candidate
+            (_, None) => None,
+
+            // No updates from the old SK in this batch, but some candidate is available:
+            // check how long time ago did we receive updates from the current SK, switch connections in case it's over the threshold
+            (None, Some((&new_sk_id, _, new_wal_producer_connstr))) => {
+                match self.wal_connection_data.as_ref() {
+                    Some(current_connection) => {
+                        let last_sk_interaction_time =
+                            match current_connection.last_wal_receiver_data.as_ref() {
+                                Some((_, data_submission_time)) => *data_submission_time,
```

Comment:
It is better to compare here not just the latest message timestamp, but timestamp of latest commit_lsn change -- otherwise we might stuck connected to safekeeper which doesn't receive any new WAL but successfully sends keepalives.